nakamoto-p2p/src/prototype.rs

```diff
@@ -0,0 +1,888 @@
+    use bitcoin::network::message_blockdata::GetHeadersMessage;
+    use bitcoin::network::message_network::VersionMessage;
+    use bitcoin::util::hash::BitcoinHash;
+
+    use nakamoto_chain::block::time::AdjustedTime;
+    use nakamoto_chain::block::tree::BlockTree;
+    use nakamoto_chain::block::{BlockHash, BlockHeader, Height};
+
+    /// User agent included in `version` messages.
+    pub const USER_AGENT: &str = "/nakamoto:0.0.0/";
+    /// Duration of inactivity before timing out a peer.
+    pub const IDLE_TIMEOUT: time::Duration = time::Duration::from_secs(60 * 5);
+    /// How long to wait between sending pings.
+    pub const PING_INTERVAL: time::Duration = time::Duration::from_secs(60);
+    /// Number of blocks out of sync we have to be to trigger an initial sync.
+    pub const SYNC_THRESHOLD: Height = 144;
+    /// Minimum number of peers to be connected to.
+    pub const PEER_CONNECTION_THRESHOLD: usize = 3;
+    /// Maximum time adjustment between network and local time (70 minutes).
+    pub const MAX_TIME_ADJUSTMENT: TimeOffset = 70 * 60;
+
+    /// A time offset, in seconds.
+    pub type TimeOffset = i64;
+
+    /// A finite-state machine that can advance one step at a time, given an input event.
+    /// Parametrized over the message type.
+    pub trait Protocol<M> {
+        /// Process the next event and advance the state-machine by one step.
+        /// Returns messages destined for peers.
+        fn step(&mut self, event: Event<M>) -> Vec<(PeerId, M)>;
+    }
+
+    /// A protocol event, parametrized over the network message type.
+    #[derive(Debug)]
+    pub enum Event<M> {
+        /// New connection with a peer.
+        Connected {
+            /// Remote peer id.
+            addr: PeerId,
+            /// Local peer id.
+            local_addr: PeerId,
+            /// Link direction.
+            link: Link,
+        },
+        /// Received a message from a remote peer.
+        Received(PeerId, M),
+        /// Sent a message to a remote peer, of the given size.
+        Sent(PeerId, usize),
+        /// Got an error trying to send or receive from the remote peer.
+        Error(PeerId, Error),
+    }
+
+    ///////////////////////////////////////////////////////////////////////////////////////////////
+
+    /// An instantiation of `Protocol`, for the Bitcoin P2P network. Parametrized over the
+    /// block-tree.
+    #[derive(Debug)]
+    pub struct Rpc<T> {
+        /// Peer states.
+        pub peers: HashMap<PeerId, Peer>,
+        /// Peer configuration.
+        pub config: Config,
+        /// Protocol state.
+        pub state: State,
+        /// Block tree.
+        pub tree: T,
+
+        /// Network-adjusted clock.
+        clock: AdjustedTime<PeerId>,
+        /// Set of connected peers that have completed the handshake.
+        connected: HashSet<PeerId>,
+        /// Set of disconnected peers.
+        disconnected: HashSet<PeerId>,
```

Comment:
It might be too early but the general architecture could benefit from the abstraction of a PeerList/Addressbook, which gives a proper API for the common queries about peers.